components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/listeners/KeyManagerUserOperationListener.java

```diff
@@ -61,32 +71,68 @@ public int getExecutionOrderId() {
     }
 
     /**
-     * Deleting user from the identity database prerequisites.
+     * Deleting user from the identity database prerequisites. Remove pending aprroval requests for the user and remove
+     * the gateway key chache.
      */
     @Override
-    public boolean doPreDeleteUser(String username, UserStoreManager userStoreManager)
-                                                                    throws UserStoreException {
+    public boolean doPreDeleteUser(String username, UserStoreManager userStoreManager) throws UserStoreException {
 
+        boolean isTenantFlowStarted = false;
+        ApiMgtDAO apiMgtDAO = new ApiMgtDAO();
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(APIUtil.replaceEmailDomainBack(username));
+
+            if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                isTenantFlowStarted = true;
+                PrivilegedCarbonContext.startTenantFlow();
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+            }
+            int tenantId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+            Tenant tenant = APIKeyMgtDataHolder.getRealmService().getTenantManager().getTenant(tenantId);
+            Map<String, String> userStoreProperties = userStoreManager.getProperties(tenant);
+            String userDomain = userStoreProperties.get(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME);
+
+            // IdentityUtil.addDomaintoName adds domain to the name only if domain name is not PRIMARY
+            // therefore domain name should be manually added to the username if domain is PRIMARY
+            if (UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME.equals(userDomain)) {
+                username = userDomain.toUpperCase() + UserCoreConstants.DOMAIN_SEPARATOR + username;
+            } else {
+                username = IdentityUtil.addDomainToName(username, userDomain);
+            }
+
+            WorkflowExecutor userSignupWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_USER_SIGNUP);
+            String workflowExtRef = apiMgtDAO.getExternalWorkflowReferenceForUserSignup(username);
+            userSignupWFExecutor.cleanUpPendingTask(workflowExtRef);
+        } catch (WorkflowException e) {
+            log.error("Error while cleaning up workflow task for the user: " + username, e);
+        } catch (APIManagementException e) {
+            log.error("Error while cleaning up workflow task for the user: " + username, e);
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
```

Comment:
Updated method signature in https://github.com/praminda/carbon-apimgt/commit/06b8f56a8eb0fa81126c41fdce3f41e90a87b5c5